List View: Displace list view items when dragging (a bit more WYSIWYG)

andrewserong opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Partly addresses: https://github.com/WordPress/gutenberg/issues/56539, follows discussion on https://github.com/WordPress/gutenberg/pull/56579.

> [!NOTE]
> 🚧 This PR is still a bit of a hack — for the moment, let's focus on the usability issues, we can tidy the code up later.

This PR explores one potential approach for displacing list view items when dragging within the list view. Or, to put it another way, while dragging blocks in the list view, move the other blocks out of the way.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The goal is to create a more WYSIWYG experience while dragging blocks in the list view. By displacing items, and using a drop indicator that effectively previews what the list item will look like when a user finished dropping their item, the list view should feel a bit more "solid" to use.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

While dragging over the list view, move existing list view items up or down to make room for a drop indicator that looks like what the block looks like once it's dropped. I.e. a drop indicator that is a bit like a preview. Details "how" follows:

* Generate an object keyed by `clientId` containing the global index for all blocks within the list view.
* In the list view branch, for each block, look up the index of the current block and compare it to the index for the block being dragged over. Move blocks out of the way by applying either an `is-above` or `is-below` classname.
* Use `transform: translateY()` rules to offset the blocks so that they move out of the way. By using CSS for this, we get nice and smooth animations which disguises the fact that the dragging behaviour only _really_ updates every `200ms`. This hopefully gives the appearance of smoother drag and drop, without (again hopefully) degrading performance.
* Hide the normal list view drop indicator line, and use a drop indicator that previews the list view item in its final position (make it more WYSIWYG)
* Auto-expand list view items on hover
* Make the drag chip used in the list view be slightly transparent so that you can see where you dropping underneath it

<!--
## Feedback / ideas requested

This approach introduces a number of challenges to be resolved. Some of the questions / things to resolve include:

* What is the experience like for nesting blocks within a collapsed block? Would it be helpful to auto-expand blocks on hover?
* How about dragging through the block hierarchy? For example, if you have a deeply nested group block, and then drag directly beneath it, but you want your block to be one or two levels higher. Dragging through the block hierarchy is potentially clearer with the drop indicator line, which this PR hides.
* Is the drag chip obscuring things too much while dragging?
* How about the timing of block moving out of the way?

-->

## To-do

* [x] Leave a block sized space when dragging starts (don't move blocks around until the user drags over another block)
* [x] Smoothly handle dealing with the space when dragging outside of the list view area (i.e. don't leave the space where it is)
* [x] Correctly animate when releasing the mouse cursor so that the dragged item smoothly goes back into position (or at least remove the wrong animation that is there at the moment)
* [x] Smoothly deal with the space left when dragging multiple blocks (i.e. can we use a CSS variable instead of hard-coded height for the above / below positioning)
* [x] Come up with a better solution for dragging into collapsed blocks
* [x] Update styling of the drag chip
* [x] Look into performance issue when dragging over editor canvas
* [x] Pick a different level of opacity / make it easier to see where you're dragging
* [x] Come up with a solution for dragging to the desired hierarchy at the end of a container block
* [x] ~Smooth dropping of list view items doesn't appear to be working in the site editor~ Known issue, unrelated to this PR: https://github.com/WordPress/gutenberg/pull/57133/files#r1444129276
* [x] Fix jumpiness in Safari due to scroll-while-dragging behaviour

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

In a post, page, or template with many blocks, try dragging up and down the list view. How does it feel to use? Also, try dragging and dropping on collapsed blocks — while hovering, the block should expand. Where do you get stuck in using this?

Try a post, page, or template with heavy nesting. How does it feel to drag to the end of a nested block, and adjust the hierarchy of where you're dropping?

## Screenshots or screencast <!-- if applicable -->

<img width="392" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/18e99d9e-0468-43f4-8e0e-1d87a43b0e29">

https://github.com/WordPress/gutenberg/assets/14988353/c102a48c-cca3-479d-939e-f8e8812a610b

<!--
<img width="377" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/62c97f6e-2c4b-422e-9b9f-c20d7bc4d453">

https://github.com/WordPress/gutenberg/assets/14988353/d2c8f98a-38e4-4043-88c9-41f508c87357
-->
<!--
<img src="https://github.com/WordPress/gutenberg/assets/14988353/16884480-0f62-4153-928b-d7d0181e5853" width="500" />

Screengrab of dragging at the root level, and then dragging to the end of a nested block, allowing a user to drag to a particular level of the block hierarchy:

https://github.com/WordPress/gutenberg/assets/14988353/1355305f-4c5e-413d-9ed1-ee2a2a700b5a
-->
<!--
https://github.com/WordPress/gutenberg/assets/14988353/3b6d0a8d-3108-4ee3-8a1b-419a855dbe6b
-->
<!--
https://github.com/WordPress/gutenberg/assets/14988353/d0111537-bca2-47b6-ad9c-7e111e900cab
-->
<!--
The following screengrab demonstrates dragging up and down the list view, while also looking at the current behaviour for nesting within a collapsed block. I think nesting within a collapsed block will need some work:
-->
<!--
https://github.com/WordPress/gutenberg/assets/14988353/0e4e0ca9-b474-4ea7-b209-8ed31877d49f

-->

andrewserong committed:
Include earlier experiment to make drag chip resemble the list view item

andrewserong committed:
Try displacing list view items via transform: translateY while dragging

andrewserong committed:
Try adding a bit of styling when nesting

andrewserong commented:
To expand on the issue of nesting within a block, if we attempt to remove the space when hovering over to perform a nesting gesture, we can get stuck. The following is an example of this if we were to remove the space (this PR does not currently remove that space):

https://github.com/WordPress/gutenberg/assets/14988353/f4ad50ac-2297-46a8-b1f6-80bb2fdc9e50

As a result, I think I'm leaning toward the idea of expanding over collapsed blocks while dragging as in https://github.com/WordPress/gutenberg/pull/56579 🤔

andrewserong committed:
Try to fix tests

talldan commented:
> When you begin to drag, the blocks move up and down to get out of the way

I'm trying to think about it in my head which is difficult, but here's what I imagine:
- If you start dragging the first block, it leaves a block-sized space, when you move down, the blocks below move up into that space
- If you start dragging the second block, it also leaves a space. If you drag up, the blocks above move down into that space, if you drag down the block below move up into that space.
- If you start dragging the last block, it again leaves a space. If you drag up, the blocks above move down into that space.

Kind of like one of those sliding puzzles 😄 

I guess that means don't move the surrounding blocks when dragging starts. Or maybe if you have to, only ever move them downwards to create the space.

jasmussen commented:
Nice! Outside of fine-tuning the drop interactions. I did have some trouble invoking the drag and drop, though, simply clicking and dragging invokes a text selection, I have to press and hold for a bit:

![dragging](https://github.com/WordPress/gutenberg/assets/1204802/76652590-6f16-46a1-8b2d-b213349838a4)

Getting the basics of this interaction right will no doubt be the majority of the headache, so I'll let you focus on that, but this seems to have great potential. If we can get it right, though, there are a couple of visual refinements that could be nice to do, such as the grabby hand cursor, the right drop shadow instead of the border, and a 2px radius on the dragged item. 

Also probably a doozy, if you are dragging an expanded tree, it would be nice to collapse that in an animated way before dragging, and then not open it on drop.

richtabor commented:
> I guess that means don't move the surrounding blocks when dragging starts. 

My first reaction as well. I wouldn't expect them to move when first invoking the drag.

apeatling commented:
> My first reaction as well. I wouldn't expect them to move when first invoking the drag.

Agreed, this was my first reaction as well. 👍

andrewserong commented:
Wonderful, that's super helpful first impressions and early feedback, folks! 🙇 I'll dig in and look at that first interaction / movement when you start dragging.

I'm also keen to hear what everyone thinks about the desired behaviour when going to nest within collapsed blocks, but the feedback already gives me heaps to work on, so absolutely no rush on re-testing at this stage 🙂

ramonjd commented:
I really like the displacement!

> What is the experience like for nesting blocks within a collapsed block? Would it be helpful to auto-expand blocks on hover?

I think it would, because:

1. I might want to place my block in a specific place within the collapsed group
2. The collapsed group might contain collapsed groups itself, into which I want to place the block. Could sub-collapsed blocks expand as well? Great if possible!
3. When I drop the block, it'd be also nice to know where in the group it's been dropped. 


https://github.com/WordPress/gutenberg/assets/6458278/8e3636c6-bec7-48ec-94a2-1a92f1adcdb8


> Dragging through the block hierarchy is potentially clearer with the drop indicator line, which this PR hides.

I do miss the line and was wondering if there could be a visual indicator of the drop zone. I just spent 30 mins trying to demo one before realizing that `:last-of-type` et. al. work on elements only.  😄 

> Is the drag chip obscuring things too much while dragging?

Not for me  personally, though I think a drop indicator would help mark the drag chip target


> How should we handle things if a browser has prefers-reduced-motion?

No transitions? I wouldn't miss them personally.

andrewserong committed:
Offset the top of the table to match the displacement

andrewserong commented:
I've pushed a tiny update, still not ready for re-testing yet, but slowly getting closer to the desired pick up and drop states. It now looks _closer_ to all the blocks moving down to create the space. Note to self: I just need to make the first transition (from starting to drag) instantaneous rather than animated to give it the appearance of creating a space where the original dragged block was. Technically, the space needs to be half "above" and half "below" so that dragging within the space is balanced between upward and downward gestures. There's a few different ways to handle that, so I'll try out a couple more next week. For reference, here's the current state of the PR (I still need to tweak / remove the animation when starting to drag, and when performing a drop):

https://github.com/WordPress/gutenberg/assets/14988353/58f58e06-185d-4a0d-bf91-24b2d9b3fca1

Thanks again for all the feedback thus far! I'll ping again once this is ready for another round of reviews.

andrewserong committed:
Improve the experience when beginning to drag a block

andrewserong commented on packages/block-editor/src/components/list-view/style.scss:
```diff
@@ -123,6 +127,35 @@
+		transform: translateY(36px) !important;
+	}
+
+	&.is-dragging {
```
Note to self: if this rule stays, add a comment explaining why it's needed (we're effectively hiding the real appearance of the dragged blocks, without `display: none` so that there's no flicker when beginning to drag)

andrewserong commented on packages/block-editor/src/components/list-view/leaf.js:
```diff
@@ -32,18 +32,18 @@ const ListViewLeaf = forwardRef(
-			enableAnimation: true,
-			triggerAnimationOnChange: path,
-		} );
+		// const animationRef = useMovingAnimation( {
+		// 	isSelected,
+		// 	adjustScrolling: false,
+		// 	enableAnimation: true,
+		// 	triggerAnimationOnChange: path,
+		// } );
 
```
This is commented out for now while I figure out the right way to handle smoothly moving the dragged item into its final position.

andrewserong committed:
Smoothly hide gap when dragging outside of the list view area

SaxonF commented:
Looking good!

richtabor commented:
> I've made some good progress on the behaviour when you initially start dragging.

Feels very nice. ✨

apeatling commented:
Looking really good @andrewserong!!

andrewserong committed:
Merge in changes from drag to collapsed block to expand PR

andrewserong committed:
Try collapsing selected blocks down to a single row space while dragging

andrewserong commented:
Thanks for the continued encouragement, folks! Another quick update: I've merged in the prototype for dragging over collapsed blocks to expand them, and I've had a go at collapsing multiple block selections down to a single row, to make it easier (and more accurate) when dragging. I'm not 100% sure on these behaviours, and it'll most likely need a fair bit of refining. Here's how those two things are currently looking:

### Multiple selection drag

I think it'd likely be a bit hard to use if we tried to have the dragged item be _all_ the blocks in the selection, so for now, I'm trying collapsing the selection down into a single row while dragging. Happy to explore other ideas if this doesn't feel nice to everyone:

https://github.com/WordPress/gutenberg/assets/14988353/e16d3719-a0f5-484e-b05f-e2cb319abd2c

### Dragging over parent to expand

I've merged in the changes from #56579 so that we can drag over collapsed blocks to expand them. I think it still needs a bit of tweaking, but it feels better to me than without the feature:

https://github.com/WordPress/gutenberg/assets/14988353/52fbb280-467a-43bc-ae94-5de9202f10f2

---

That's it for today. I have a bit more to do before switching this over to "ready for review" — namely correctly animating the drop action when a user releases the mouse button. I think at this stage that's probably the last thing for me to implement before moving this out of draft. As always, happy for any feedback or ideas along the way, though 🙂

tellthemachines commented:
Just had a little play with this and it's very smooth. I especially love that we're now able to drop into closed containers!

One thing I noticed is that if dragging multiple blocks the space where they were doesn't close up like it does with a single block:

https://github.com/WordPress/gutenberg/assets/8096000/bbed8c07-fa7f-4950-a161-67eb505a5302

Is this intended? It's not terrible either way (I'm thinking if one regrets the action half way through, the space closing up means it's harder to find the previous position, but there's always undo for that so it's probably not an issue 😅 )

andrewserong commented:
Thanks for testing!

> One thing I noticed is that if dragging multiple blocks the space where they were doesn't close up like it does with a single block:

Hrm, they should be closing up for a multi-selection, as the `.is-after-dragged-blocks` classname should be handling collapsing the area of the multiple selection when dragging beneath it. Do you mind re-testing with a couple of different posts / also in the site editor, and see if it's happening consistently for you? I haven't been able to replicate that so far 🤔

> How do you make a different between dropping at the end of a container and dropping after the container. It seems that it might be a bit confusing visually to make the distinction?

That's a really good point. I wonder if we could add in a (perhaps subtle) vertical drop indicator line to indicate nesting level when at the end of a container block. I'll have a play around.

tellthemachines commented:
> Do you mind re-testing with a couple of different posts / also in the site editor, and see if it's happening consistently for you?

I couldn't reproduce it again, so might have been some env issue 🤷 it's working fine for me now!


https://github.com/WordPress/gutenberg/assets/8096000/3b893bf8-adfb-42d8-9128-a0d144135baf

andrewserong committed:
Fix jumpiness in Safari

andrewserong committed:
Try making styling match the selected state

andrewserong commented:
Small styling update: in order to come up with styling that will work for both light and dark modes of the list view, I thought I'd try styling the dragged component as though it's selected. I think it potentially helps give it a bit more consistency, while avoiding the problem of needing to come up with multiple sets of styling rules for the dragged element depending on where the list view is used. Here's a couple of screenshots (and happy for feedback or other ideas, of course!):

| Dark view list view | Light view list view |
| --- | --- |
| ![image](https://github.com/WordPress/gutenberg/assets/14988353/937deb20-ed8d-49d2-b891-c6664bd5be6f) | ![image](https://github.com/WordPress/gutenberg/assets/14988353/f70a28d2-dde6-48c8-bc1a-215fa0edb9c4) |

andrewserong committed:
Try smoothly dropping to the final position

andrewserong commented on packages/components/src/draggable/index.tsx:
```diff
@@ -160,6 +160,7 @@ export function Draggable( {
 
 			const clone = element.cloneNode( true ) as HTMLElement;
 			clone.id = `clone-${ elementId }`;
+			clone.classList.add( 'components-draggable-cloned-element' );
```
I couldn't quite work out a good alternative for allowing the `leaf` items to look up (or pass in a reference to) the _one_ cloned drag component that is currently in use. Adding a classname seemed a simple(ish) way to do it, but it does mean that the classname would become part of this component's API.

apeatling commented:
Gave it a spin, it's feeling really nice, and is a huge improvement already. Few bits of feedback:

With the solid styling on dragging, I find it more frustrating to confirm I'm dragging on top of the right group because I can't see it behind what I'm dragging. I think some level of reduced opacity is better.

<img width="385" alt="Screenshot 2023-12-08 at 8 55 17 AM" src="https://github.com/WordPress/gutenberg/assets/1464705/65c05a30-6384-46e8-9c5c-d27e41de067e">

When I drag a child above and below an item, the animation works for the item sliding down, but not back up, it snaps back.

https://github.com/WordPress/gutenberg/assets/1464705/e588ec60-bc5d-4088-b8b3-2e94db23603f

It still feels a bit too hard to stop on a parent element to let it expand, it's really easy to trigger the item to move out of the way, or for the animation to freak out a bit. I wonder if we can make the drop area bigger where it doesn't trigger either the above or below action?

https://github.com/WordPress/gutenberg/assets/1464705/e7189755-c013-4419-b303-2e34f24a901e

apeatling reviewed (commented):
Gave it a spin, it's feeling really nice, and is a huge improvement already. Few bits of feedback:

With the solid styling on dragging, I find it more frustrating to confirm I'm dragging on top of the right group because I can't see it behind what I'm dragging. I think some level of reduced opacity is better.

<img width="385" alt="Screenshot 2023-12-08 at 8 55 17 AM" src="https://github.com/WordPress/gutenberg/assets/1464705/65c05a30-6384-46e8-9c5c-d27e41de067e">

When I drag a child above and below an item, the animation works for the item sliding down, but not back up, it snaps back.

https://github.com/WordPress/gutenberg/assets/1464705/e588ec60-bc5d-4088-b8b3-2e94db23603f

It still feels a bit too hard to stop on a parent element to let it expand, it's really easy to trigger the item to move out of the way, or for the animation to freak out a bit. I wonder if we can make the drop area bigger where it doesn't trigger either the above or below action?

https://github.com/WordPress/gutenberg/assets/1464705/e7189755-c013-4419-b303-2e34f24a901e

andrewserong committed:
Fix transitions above dragged items

andrewserong commented:
Thanks for highlighting all those details @apeatling! I've pushed a partial fix for some of these issues in https://github.com/WordPress/gutenberg/pull/56625/commits/2d532fc366ceee5d1216701b21752a57340448d5 — for Safari in particular, but also in some cases in Chrome, we need to add a separate class for the "normal" state so that the blocks above those being dragged will smoothly animate back into position, while not affecting the drop animation. That part appears to be largely fixed, now, I think.

I noticed that there can be a bit of jumpiness in Safari that appears to be to do with the existing `useScrollWhenDragging` behaviour:

https://github.com/WordPress/gutenberg/blob/68ca9954a5c8b736fe00fc7ada30c3b3cee9675a/packages/block-editor/src/components/block-draggable/index.js#L53

I _think_ what's happening there is that the scroll on drag behaviour is active even when there isn't a scrollbar in the list view. The buggy / jumpy behaviour there is likely (technically) present on `trunk`, but more exposed in this PR due to the moving around we're doing of the list view elements. I'll continue digging here.

> It still feels a bit too hard to stop on a parent element to let it expand, it's really easy to trigger the item to move out of the way, or for the animation to freak out a bit. I wonder if we can make the drop area bigger where it doesn't trigger either the above or below action?

I think this issue should be a bit more improved now that the "normal" state class is present, as it should have reduced some of the jumpiness. I'm sure this could be tweaked further, so happy for more feedback on the ideal behaviour here, too — it's a tricky thing to find the right balance when dragging over a container block, because it's hard to know if the user is attempting to nest or attempting to drag past it. Some of that can be tweaked by adjusting the time before we expand the collapsed block, too.

> With the solid styling on dragging, I find it more frustrating to confirm I'm dragging on top of the right group because I can't see it behind what I'm dragging. I think some level of reduced opacity is better.

Same, I'll have a play around with reduced opacity, but it'd be great to get some feedback from designers on an ideal look for the dragged state when folks are free! (No rush of course, there's still a fair bit to improve on here as-is). If we go with any colour other than the selected state, then we'll need to factor in both light and dark modes as the list view is also used in the Navigation section of the site editor.

andrewserong commented:
Note to self: there also appears to be a performance issue when dragging over the editor canvas, so there's likely something happening too frequently in this PR / an introduction of an expensive operation somewhere. I'll add it to the task list.

andrewserong commented:
Update: just gave this a rebase and fixed the merge conflicts so that the PR is current again. I still have a few more bits of jankiness to resolve before this'll be ready for a proper review, but the to-do list in the PR description is current (with the issues I'm aware of). I'll continue on with this PR in the new year!

andrewserong commented:
Note to self from debugging the jumpiness in Safari that isn't present in Chrome: If I log out the throttled `onDragOver` callback in the list view in Safari, it's being called much more frequently than the throttle. It's most obvious when I [update the throttle to `1000ms`](https://github.com/WordPress/gutenberg/blob/9dc85d8748e3086520b3c66714902494062b3956/packages/block-editor/src/components/list-view/use-list-view-drop-zone.js#L553) and then drag within the list view:

https://github.com/WordPress/gutenberg/assets/14988353/3d5f95e2-f905-46b4-9def-b75f46676c5a

The same applied in Chrome correctly only updates once a second:

https://github.com/WordPress/gutenberg/assets/14988353/d7e1646d-9d1f-4c41-93b6-05fa44f1d13b

I'll do some more digging, but this gives a solid lead on the cause now. The ideal state is that the throttled function's `wait` time should be roughly the same as the transition of the animation, so that the animation is complete (or close to it) before we try to move things out of the way again.

andrewserong committed:
Fix flickering issue in Safari

andrewserong commented:
Progress on the Safari jumpiness issue: it turns out it's the same type of problem fixed separately for the `DropZoneComponent` back in #44264. It's due to a bug in Safari (https://bugs.webkit.org/show_bug.cgi?id=66547) that means that our `onDropLeave` callbacks fire more frequently than expected (in Safari, it'll fire when dragging over children of the drop zone). The workaround (borrowed from #44264) is to apply `pointerEvents: none` while dragging blocks, so that `onDragLeave` isn't fired while dragging within the list view drop zone area. I've done that here in https://github.com/WordPress/gutenberg/pull/56625/commits/162c1f179f36c0299b9a90772ff9109c49003b5f, and that seems to have resolved the problem in Safari now (the problem being that `onDragLeave` [cancels out the throttle](https://github.com/WordPress/gutenberg/blob/9dc85d8748e3086520b3c66714902494062b3956/packages/block-editor/src/components/list-view/use-list-view-drop-zone.js#L564) when a user drags outside of the list view, but this was happening within the list view).

Here's how it's looking in Safari now — much smoother:

https://github.com/WordPress/gutenberg/assets/14988353/c179dc25-6aa4-40b9-a4d8-8cc31606ca7b

Next up to work on:

* [ ] Pick a different level of opacity / make it easier to see where you're dragging
* [ ] Come up with a solution for dragging to the desired hierarchy at the end of a container block
* [ ] Smooth dropping of list view items doesn't appear to be working in the site editor
* [ ] Tidy up the code

Getting much closer to being ready for another round of reviews. I'll continue working on this next week and ping for reviews and further feedback, then!

andrewserong committed:
Try a transparent background color

andrewserong committed:
Fix drag behaviour in navigation block

andrewserong commented:
Small update:

* I've resolved a regression with the Navigation block caused by the Safari fix. The solution was to move the `pointer-events: none` rule down to the `tbody` element so that drag behaviour is still recognised on the container, while preserving the Safari fix.
* I've switched to a transparent background color for the drag chip — this makes it considerably easier to see what you're doing, but can obviously tweak this further.
* The source of the issue where the dropped list view item sometimes wasn't animating into position was unrelated to this PR. It turns out `useMovingAnimation` switches off when there's 200+ blocks in the document. I've asked a question on the PR that introduced that to see if we can exclude the list view from that logic: https://github.com/WordPress/gutenberg/pull/57133/files#r1444129276

Here's how it's all currently looking:

https://github.com/WordPress/gutenberg/assets/14988353/ddc81600-69f9-4a30-8c0f-88b006cbef65

Thinking about this item a little more:

> Come up with a solution for dragging to the desired hierarchy at the end of a container block

If we go with a transparent background color for the drag chip (as is now implemented in this PR), then we have a bit more wiggle room to play with to make it clearer where we're dropping within the desired hierarchy. The next thing I'd like to play around with is to try out a small vertical drop indicator line that only gets displayed when we're changing the hierarchy at the end of a group block. Hopefully a little visual indicator there might resolve this item.

andrewserong committed:
Add vertical drop indicator

andrewserong commented:
Update: I've had a go at implementing a vertical drop indicator line that shows which level of the block hierarchy the user will be dropping to. That, combined with a transparent background on the drag element, and I think we might be honing in on something that can provide the desired WYSIWYG of "really" dragging the list view item, while preserving the specificity of a drop indicator line. Here's a screenshot of how it's looking now:

<img src="https://github.com/WordPress/gutenberg/assets/14988353/1d2b2442-67ec-47f4-af0a-23e7d4d8ef50" width="500" />

And a screengrab:

https://github.com/WordPress/gutenberg/assets/14988353/1355305f-4c5e-413d-9ed1-ee2a2a700b5a

It's challenging to find a background color that works for both light and dark UIs (it'd be great if we don't have to have two sets of styling rules). Here's how it's looking in the navigation part of the site editor:

https://github.com/WordPress/gutenberg/assets/14988353/40224edb-83da-43c8-8f29-a63b4942d96d

---

I've now worked through the main outstanding issues that I was aware of. There's still a little bit of inconsistency with when the smooth dragging into the final position applies and when it doesn't. Some of that might need to be looked into separately to this PR as it gets into the weeds of `useMovingAnimation` a little, and how the rendering of the list view works. I'll switch this over to "Ready for review" now.

## Next steps

The next steps for this PR is that I believe we're now ready for design feedback on the look and feel of the dragged list view item and the presentation of the vertical drop indicator line (and whether folks like the existence of the latter). For context: the vertical drop indicator line exists so that users can see where they're dropping to at the end of a nested block or template. Without it, it feels hard to use to determine what's going to happen when you let go of the mouse button.

Happy for any and all feedback, especially on how it feels to use! The code is still a bit messy, but I'd like to address all UX / UI issues before finalising everything.

talldan commented:
Looks great! I've also seen some examples of other apps (can't remember where though) where the draggable chip snaps horizontally to the level of  indentation, so that could also be an alternative approach to try.

ramonjd commented:
As far as functionality goes, this is working excellently for me! For me the line indicator appears where it should, with helpful indentation.

> It's challenging to find a background color that works for both light and dark UIs

I forgot, was there a requirement that the dragging element be opaque? To make it easier to see a group list item underneath or something, right?

I'd be interested to hear feedback about the line.  I'm wondering if a more solid line might be clearer?

https://github.com/WordPress/gutenberg/assets/6458278/be149985-8216-4c11-810f-381b9db7c5f0

I expect it wouldn't straightforward to style unfortunately.

andrewserong commented:
Thanks for testing, folks! 🙇

> I forgot, was there a requirement that the dragging element be opaque? To make it easier to see a group list item underneath or something, right?

The main thing we've found in earlier testing in this PR is that the dragged item needs to be transparent (or a non-opaque background color) so that we can see where we're nesting, especially at the end of a Group block or Template where the user might be dropping in a number of different available positions of hierarchy.

> I'd be interested to hear feedback about the line. I'm wondering if a more solid line might be clearer?

Thanks for trying out another styling idea! Whichever color or size for the drop indicator we go with (if we use one), it'll likely work best if it can be a color with an opacity set, so that it'll play nicely on light or dark backgrounds. I quite like the idea of showing a whole area for where it'll drop like in your screengrab, and it's less intrusive than a single vertical line, while still signalling nesting level. If it's bigger like in your screengrab, I'd probably go for a slightly more subtle color / a fair amount of opacity.

Let's see what the designers think of our options! 🙂

jameskoster commented:
Is the drop indicator the same height as other rows to avoid 'reflow' on drag? It looks a bit heavy and I'm wondering if the current pattern (thin blue line) might work instead?

<img width="304" alt="Screenshot 2024-01-09 at 10 04 37" src="https://github.com/WordPress/gutenberg/assets/846565/7a278d4c-3072-42ec-b7e5-709ccc26efe5">

andrewserong committed:
Try using existing drop indicator line

andrewserong commented:
> Is the drop indicator the same height as other rows to avoid 'reflow' on drag? It looks a bit heavy and I'm wondering if the current pattern (thin blue line) might work instead?

At this stage, the size and shape of the drop indicator is fair game to try anything out! My main concern is ensuring that folks can clearly see which level of nesting they're going to be dropping at before they let go of the mouse cursor. But we can try out many different approaches to that.

It's an interesting idea to see what it looks like with the current pattern of the thin blue line. It needed a little bit of tweaking, but here's how it's looking now:

<img width="397" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/4a5fa93a-a8f3-4766-9077-7bce6f6b6742">

And a screengrab:

https://github.com/WordPress/gutenberg/assets/14988353/84244bc7-9ac8-44a6-973f-6eae758a3491

What do folks think? The drop indicator line is perhaps a bit sudden in contrast to the smooth animation of list view items moving up and down. But perhaps we could try animating the show/hide of the drop indicator line in some way as flagged in #39068 by @richtabor 🤔

andrewserong committed:
Add reduced motion, fix e2e test

talldan commented:
I personally find the horizontal line in the video visually confusing, maybe if it was centered in the empty space it might be clearer, though it could also interfere with the draggable chip. Hard to say without seeing it in action.

andrewserong commented:
> maybe if it was centered in the empty space it might be clearer, though it could also interfere with the draggable chip. 

Yeah, it does interfere with the draggable chip a bit if we try centering it. That doesn't feel quite right to me, but it was worth a try at least! Here's a screengrab of that idea (the state of this PR still has it underneath):

https://github.com/WordPress/gutenberg/assets/14988353/b3d43624-18a3-4aff-a69f-f8f51d10c2f9

To recap on some of the explored or potential options so far:

* A vertical drop indicator line that is `4px` wide and the height of the list view item space, to show which level of nesting we're in, or:
* Use the existing drop indicator line as on trunk, but positioned at the bottom of the empty space, or
* Use the existing drop indicator line as on trunk, but positioned at the middle of the empty space, or
* Display a box that is the shape of where the list view item will be dropped, but faded out (shows the rough size and shape of where the user will drop)

@jameskoster @jasmussen @SaxonF @richtabor do any of you want to have a go at a visual mockup of what an ideal drop indicator could look like that plays nicely with the displacement in this PR? (Or alternately, do you have any strong opinions on any of the ideas above 🙂)

talldan commented:
Looking around for other ideas, and the demos here keeps the block/item in place rather than leaving a blank space - https://sortablejs.github.io/Sortable/

That could be something else to try. 😄 

Though I also think it's fine to land whatever feels best out of those current versions (the box is the one I like personally).

jameskoster commented:
It's tricky to get the design right for this as it really needs to be 'felt' live to get a good idea about what works. But I agree with @talldan that the 'box' approach works best out of the options tried so far. 

A nice detail about the Sortable example is that the nesting level can be communicated by making the content visible inside the drop indicator. Perhaps we try something in between; the 'box' approach with the contents visible inside:

<img width="2404" alt="Drag" src="https://github.com/WordPress/gutenberg/assets/846565/51cca421-7825-4ebc-9018-1bebb3c061dd">

Another tweak we might try is removing the nesting indication in the chip while being dragged as that detail becomes variable once the block is 'picked up'.

andrewserong commented:
Wonderful, thank you both! That gives me some good ideas to start playing with, and we can see how it feels (and how far we can get) 🙂

andrewserong committed:
Tidy up classname for drag chip in prep for subsequent changes

andrewserong committed:
Remove indent from drop chip, ensure indent is applied to the position

andrewserong committed:
Try updating the drag chip to use a border

andrewserong committed:
Try drop indicator line as preview of final list item

andrewserong commented:
Update: this feels pretty promising to me! I've had a go at implement the ideas from @jameskoster in this comment: https://github.com/WordPress/gutenberg/pull/56625#issuecomment-1884709624

Here's how it's looking:

https://github.com/WordPress/gutenberg/assets/14988353/d2c8f98a-38e4-4043-88c9-41f508c87357

I don't think it looks _terrible_ on the dark background, but the white background color is a little jarring. Perhaps we can tweak the opacity a bit to find a sweet spot that works for both environments (it's a bit unfortunate as `0.5` opacity looks better to my eyes on a white background, but much lower like `0.1` on a dark background). Here's how it's currently looking in a dark mode:

https://github.com/WordPress/gutenberg/assets/14988353/9d533050-9152-4547-8abc-09f2d332c001

The updates above include:

* Removed indent on the drag chip (that resolves a fair bit of the visual problems with deeper nested blocks, thanks for the idea, James!)
* Tweaked styling of the drag chip to include a border
* Updated the drop indicator to be effectively a "preview" of how the list item will finally look, by making it full height, with a subtle background color, and getting it to render the block icon and block title. I've added a separate component for this to make it easy to hack around in this PR without getting stuck. Once we've settled on an approach, I'll merge it in with the current drop indicator code.

It's mostly working, there's a bug with the indentation of the previewed block list item if you're in a really deep list and are scrolled to the right horizontally. It's a little fiddly to fix, so I'll look into that if we're happy with the overall direction of this approach.

What do folks think?

talldan commented:
> What do folks think?

This looks great! ❤️

ramonjd commented:
> What do folks think?

This is amazing!

jameskoster commented:
Thanks for putting that together @andrewserong. Funnily enough, now that the drop indicator clearly communicates the position and hierarchy along with the contents, the chip feels a bit superfluous, it's even a bit visually distracting. I'm left wondering whether we need it, or if we should try a more reduced chip (similar to the one currently shipping). What do you think?

One other detail: the drop indicator is harder to distinguish when dropping inside a container due to the matching backgrounds. Perhaps we could fix that by using the same background treatment: `rgba(var(--wp-admin-theme-color--rgb), .04)`?

andrewserong commented:
> I'm left wondering whether we need it, or if we should try a more reduced chip (similar to the one currently shipping). What do you think?

That's a good question. I quite like the idea for another reason, which is that it could potentially help us to reduce the scope of this PR if we're not also changing the drag chip at the same time. However, it introduces an issue for us when dragging to the different nesting levels, as the chip obscures the preview of where the block will land:

https://github.com/WordPress/gutenberg/assets/14988353/d9f83e6d-0d55-46d5-a0dd-b81bded3a79e

However, we could potentially work around that if we were happy to go with some transparency on the current drag chip:

https://github.com/WordPress/gutenberg/assets/14988353/c043d56e-3e08-4708-9146-20d2280b08e9

What do you think? Overall, I don't mind the idea of seeing if we can land this PR with the current drag chip, if we accept some level of transparency. For now, I haven't pushed any further changes to this PR (the drag chip is still the one that resembles the list view item). But if we like that idea I'll push some changes on Monday to restore the drag chip.

For now, since I think we all agree on the drop indicator line preview, I'll look into stabilising the behaviour there (fixing up the behaviour when the user has scrolled horizontally).

andrewserong commented:
Update: for ease of testing, and so that folks can try out both versions of this PR, I've left this PR using the drag chip that resembles the list view item. I've also now opened up a variant PR over in https://github.com/WordPress/gutenberg/pull/57782 that demoes using the drag chip as on trunk, so that folks can play with that if they like. Happy for any feedback / preferences that folks might have!

> One other detail: the drop indicator is harder to distinguish when dropping inside a container due to the matching backgrounds. Perhaps we could fix that by using the same background treatment: `rgba(var(--wp-admin-theme-color--rgb), .04)`?

I forgot to reply to this earlier. We're actually using those styles for the drop indicator line. The problem is that the background colors that get applied in selected nested lists are applied on each individual row, and not on the whole section, so we don't get a stacking effect of layered background colors. Perhaps one approach we could explore could be to check if any of the parents of the block we're dragging onto is selected, then add a classname that gives us a darker color? I'll have a play around.

andrewserong committed:
Try offsets for nesting level of drop indicator, and use darker color if adjacent block has selected branch

andrewserong commented:
Final update for the week: I've improved some of the offsetting logic, but I still need to fix it up properly in `rtl` languages. Also, I've added in an extra classname to apply a slightly darker background color when the dragged block is adjacent to a block with `is-branch-selected`. That appears to improve the visibility a bit:

<img width="601" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/4b07de7a-ee14-49a5-a0c7-fa5e46456c00">

https://github.com/WordPress/gutenberg/assets/14988353/e507cf00-77d4-4489-95b3-3c2ce5b84aab

I'll continue working on all this next week. It feels like we're getting closer! Thanks again for all the help, everyone 🙇

jameskoster commented:
> Also, I've added in an extra classname to apply a slightly darker background color when the dragged block is adjacent to a block with is-branch-selected

Thanks, that seems to work pretty well. 

In terms of the drag chip, I have a slight preference for the currently-shipping version (made slightly transparent), this one:

<img width="386" alt="Screenshot 2024-01-12 at 14 17 09" src="https://github.com/WordPress/gutenberg/assets/846565/a7428677-f726-4f8a-977b-d451797cc4b7">

It interferes less with the drop indicator.

But let's see what other designers say.

andrewserong commented:
Thanks for the continued feedback @jameskoster! I've updated this PR to use the drag chip as on trunk, but with an opacity of `0.8` within the list view. Here's how it's looking:

https://github.com/WordPress/gutenberg/assets/14988353/c102a48c-cca3-479d-939e-f8e8812a610b

And in RTL languages it appears to be working well, too:

https://github.com/WordPress/gutenberg/assets/14988353/a8844783-58e9-4be1-a463-ae121a159f18

This is ready for another round of design feedback / to see if we're happy with the current state and look and feel. Thanks! If / when we're happy with how it's looking, I'll tidy up the code a bit more (consolidate the drop indicator code), and get it ready for a final code review.

andrewserong committed:
Use drag chip as on trunk, but with opacity of 0.8

andrewserong committed:
Consolidate drop indicator logic

talldan commented:
Something to consider is how dragging files into list view will work. In trunk you can drag media files from explorer into list view and create the relevant blocks (though it doesn't seem to work properly when dragging into a group, that might be a bug).

In this PR I don't think it's quite working as expected, the insertion point doesn't seem to be clearly highlighted. Hopefully there's a reasonable solution for it.

andrewserong commented:
> In this PR I don't think it's quite working as expected, the insertion point doesn't seem to be clearly highlighted. Hopefully there's a reasonable solution for it.

Oh, great catch! I'll look into it.

andrewserong committed:
Fix drop indicator when dragging files

andrewserong committed:
Don't add nesting classname when dragging files as the editor thinks we can nest for any block

andrewserong commented:
Thanks again, Dan. I've updated the logic to allow the displacement when dragging files. Since we don't _really_ know which block will wind up being used while dragging, I've gone for a simple approach of just leaving a space without the block preview text. I think it seems a reasonable approach to go with for now:

https://github.com/WordPress/gutenberg/assets/14988353/c84df48a-ff93-40df-b211-82c32f39c55c

> In trunk you can drag media files from explorer/finder into list view and create the relevant blocks (though it doesn't seem to work properly when dragging into a group, that might be a bug).

I noticed that dropping files within the list view appears to randomly not work sometimes both in this PR and on trunk. I think it's most likely a bug to look into separately.

---

Let me know if anyone runs into any other UX issues. Since it feels like we're honing in on a decent shape for this, my next step will be to take some of the logic introduced in `branch.js` and see if I can move it to a function and add some tests, since the logic is likely pretty hard to read. I'll see what else I can do to tidy this PR up a little for code review.

andrewserong committed:
Move displacement logic to a helper function, add tests

andrewserong committed:
Remove unneeded changes from useMovingAnimation

andrewserong committed:
Add explanatory comments

andrewserong commented:
Update: I've moved the logic that determines when a row should be displaced up or down into a helper function and added some unit tests to cover all of the logic that's been introduced there. I think that's probably the most unreadable part of this code change, so hopefully between the unit tests, the abstracted function, and the inline comments, that helps stabilise the displacement logic somewhat.

Since there hasn't been any negative feedback on the latest direction using the drop indicator on trunk, I think this is now ready for a final round of code and/or design UX reviews. Please let me know if there's anything unclear here — this PR has been through a number of iterations, so it's highly possible I might have missed a comment here or there 😄

talldan commented:
> I've gone for a simple approach of just leaving a space without the block preview text

That looks like a good way to solve the problem while still keeping the experience consistent with dragging blocks 👍

jasmussen commented:
Sorry for not having had the time to look sooner, but this is feeling pretty good overall! GIF showing me dragging some items in the site editor:

![displacement](https://github.com/WordPress/gutenberg/assets/1204802/1aa2111b-e4c9-456b-893b-db494d7a0856)

    That said, and sorry for being the contrarian here, I think the main value of immediate displacement is that we can _get rid_ of the chip. I understand that the list item has to indent itself and snap in place, so it's never going to be perfectly in sync with where your cursor is, I sense that might be why it still exists, but it doesn't feel entirely right to me, and honestly earlier iterations of this branch felt better.

I see some conversations around dropzones etc — if need be, can we move towards something that's a bit more like Google Keep?

![keep](https://github.com/WordPress/gutenberg/assets/1204802/05d52575-c8e5-4023-9aa7-8328b1c7d434)

That still has the displacement, the thing moving in the canvas is a subtle drop-zone "slot", and the thing you are dragging is the row itself, not the chip?

jameskoster commented:
Isn't the chip necessary when dragging from List View to the Canvas? Otherwise I agree it would be nice to get rid of it. Perhaps the chip only appears when the cursor is above the canvas?

jasmussen commented:
> Isn't the chip necessary when dragging from List View to the Canvas?

Right, good point. I wonder how common that is. Could/should it transform into a chip when moving out of the list view? Could it be a "the chip is there" but it's unhidden when moving into the canvas?

andrewserong commented:
Thanks for the feedback, folks! I also really liked how it looked when the drag chip resembled the list view item, but it also raised a few styling concerns to address. With that change removed from this PR, it's made the changeset a bit easier to manage.

What do you both think about seeing if we can land this particular PR with the drag chip as it is currently, and then I can open up a separate follow-up PR to explore making it resemble the row items again, and also look into the behaviour when dragging from the list view over the editor canvas?

My main thinking here is that this PR now handles the displacement in a fairly reliable way, so it could be good to land the bits that we're pretty happy with, and then it should make it a bit easier to review if we're looking at the drag chip in isolation in a separate PR.

jasmussen commented:
Good point, happy to land a version of this. That said, I'm not seeing this work in the post editor at all, I wonder is this only on my end?

andrewserong commented:
Thanks for confirming, Joen!

> That said, I'm not seeing this work in the post editor at all, I wonder is this only on my end?

It should be working just fine in the post editor (and any instance of the list view, such as in the Navigation block or editing Navigation in the site editor, too). What were you seeing?

Here's a quick link to test this PR in Playground in case that helps rule out anything in your local env: https://playground.wordpress.net/gutenberg.html?pr=56625

jasmussen commented:
Oh that is a _nice_ PR viewer. I don't recall who owns gutenberg.run, but I wonder if we can suggest redirecting to the playground for it? This seems to work better.

That said, some hiccups which show in this GIF, some CSS appears to not be loading, and there's a missing post title placeholder. This was addressed after a refresh, but noting for posterity. And the drag and drop does appear to be working as intended:

![state](https://github.com/WordPress/gutenberg/assets/1204802/0de87382-2242-475d-8d27-270458d8eab5)

It does appear to be _slightly_ laggy. If I click and drag very quickly, it doesn't drag at all. Not world ending, but I wonder if there are some small things we can do?

![slightly laggy](https://github.com/WordPress/gutenberg/assets/1204802/9086ec42-6d6b-443c-a59c-47d43ea43c68)

It's also running in my local branch now, as intended. Not sure what was wrong when I tested yesterday, perhaps the `npm install && npm run dev` helped:

![local](https://github.com/WordPress/gutenberg/assets/1204802/78f72199-f34f-4e23-a7ed-13f074072093)

Mainly to get this landed, is my question around the lagginess. It can feel a bit choppy to move down the list of items. Are there any options here?

andrewserong committed:
Try a timeout for the list view drop zone updates of 50ms instead of 200ms

jasmussen reviewed (approved):
I'm happy to move forward with this one. Impressive work!

andrewserong commented:
> I'm happy to move forward with this one. Impressive work!

Thanks for the approving review @jasmussen! I'll merge this in now, and happy to tweak any peculiarities in follow-ups. Let me know if anyone runs into any further issues or things that need tending to.

First up, I'll look at opening a WIP PR to explore re-adding the WYSIWYG drag chip of the list view row 🙂

andrewserong commented:
Update: for anyone following along, I've opened up a follow-up PR that explores the WYSIWYG drag chip that was in earlier commits in this PR. The WIP PR is over in: https://github.com/WordPress/gutenberg/pull/58103

It's mostly working, but I'll need to do a little more tweaking before it's ready for review. One of the main issues design-wise is coming up with a background color for the drag chip that can work in both light and dark modes 🤔
